### DISCUSSION TITLE
Can I use ErrorProne to create custom annotation processor without adding it to ErrorProne?

### DISCUSSION THREAD

**Discussion by xingyutangyuan** [google/error-prone#4190] [Q&A] [answered]
The [Writing a check](https://github.com/google/error-prone/wiki/Writing-a-check) page outlines how to create a custom check into ErrorProne.

Is it possible to build a check specialized for a particular 3rd-party class (e.g. to prevent misuse), and package it up as a separate annotation processor without adding it as part of core ErrorProne?

**Comment by tbroyer** [comment, marked as answer]
You'll want to build a "plugin check" then, see https://errorprone.info/docs/plugins

**Comment by xingyutangyuan** [comment]
Thanks for the pointer!

I was able to get a basic checker plugin, but third-party dependency doesn't work.

`mvn compile` is happy with the third-party dep. But when running the test with `CompilationTestHelper`, the test failed with runtime error not recognizing any of the third-party classes.

It didn't complain about Guava classes though (and I'm setting Guava and the third-party dependency the same way in the pom file).

That makes me wonder if Error-Prone (or `CompilationTestHelper`) directly pulls in Guava as a sanctioned dependency?

**Reply by cushon** [reply]
By default `CompilationTestHelper` runs test compilations against whatever dependencies the test has. (There are ways to customize the classpath for the test compilation, see e.g. [this API](https://github.com/google/error-prone/blob/5395af43d7cf270be22257bc42d71ca93f927c87/test_helpers/src/main/java/com/google/errorprone/CompilationTestHelper.java#L216)).

So if Guava is a dependency of the test using `CompilationTestHelper`, test compilations in the individual test cases will be able to reference Guava. If the compilations in the test cases require additional dependencies, the easiest solution is to add them as dependencies of the test.

**Comment by xingyutangyuan** [comment]
The tests pass now. 

Though the annotation processor doesn't seem to do anything.

Here's how I add it as annotation processor in the pom file:

```
            <annotationProcessorPaths>
              <path>
                <groupId>com.google.errorprone</groupId>
                <artifactId>error_prone_core</artifactId>
                <version>2.23.0</version>
              </path>
              <path>
                <groupId>${project.groupId}</groupId>
                <artifactId>my-custom-check</artifactId>
                <version>${project.version}</version>
              </path>
            </annotationProcessorPaths>
```

But the check is not detecting any error.

Is there any configuration I need to do within the `my-custom-check` artifact? I already have the `auto-service` as a dependency:

```
    <dependency>
      <groupId>com.google.auto.service</groupId>
      <artifactId>auto-service</artifactId>
      <version>1.1.1</version>
    </dependency>
```

**Comment by cushon** [comment]
> Is there any configuration I need to do within the my-custom-check artifact? I already have the auto-service as a dependency:

Do you have AutoService configured on the `<annotationProcessorPath>` of the my-custom-check artifact? See e.g. https://github.com/google/auto/blob/main/service/README.md#getting-started

**Reply by xingyutangyuan** [reply]
Yeah. I should have added AutoService in the annotation processor path of the custom check, not of the "user" code.
